pg-extend/src/pg_datum.rs

```diff
@@ -370,6 +363,101 @@ where
+impl PgPrimitiveDatum for i32 {}
+impl PgPrimitiveDatum for i64 {}
+impl PgPrimitiveDatum for f32 {}
+impl PgPrimitiveDatum for f64 {}
+
+impl<'s, T> TryFromPgDatum<'s> for &[T]
+where
+    T: 's + TryFromPgDatum<'s> + PgPrimitiveDatum,
+{
+    fn try_from<'mc>(_: &'mc PgAllocator, datum: PgDatum<'mc>) -> Result<Self, &'static str>
+    where
+        Self: 's,
+        'mc: 's,
+    {
+        if let Some(datum) = datum.0 {
+            unsafe {
+                let datum = datum as *mut pg_sys::varlena;
+                if datum.is_null() {
+                    return Err("datum was NULL");
+                }
+
+                #[allow(clippy::cast_ptr_alignment)]
+                let arr_type = pg_sys::pg_detoast_datum(datum) as *mut pg_sys::ArrayType;
+
+                if (*arr_type).ndim > 1 {
+                    return Err("argument must be empty or one-dimensional array");
+                }
+
+                let mut elmlen: pg_sys::int16 = 0;
+                let mut elmbyval = pgbool!(false);
+                let mut elmalign: ::std::os::raw::c_char = 0;
+
+                pg_sys::get_typlenbyvalalign(
+                    (*arr_type).elemtype,
+                    &mut elmlen,
+                    &mut elmbyval,
+                    &mut elmalign,
+                );
+
+                let mut nulls = std::ptr::null_mut::<pg_sys::bool_>();
+                let mut elements = std::ptr::null_mut::<Datum>();
+                let mut nelems: i32 = 0;
+
+                pg_sys::deconstruct_array(
+                    arr_type,
+                    (*arr_type).elemtype,
+                    elmlen as i32,
+                    elmbyval,
+                    elmalign,
+                    &mut elements,
+                    &mut nulls,
+                    &mut nelems,
+                );
+
+                let datums = std::slice::from_raw_parts(elements as *const Datum, nelems as usize);
+
+                // This is where the conversion from `&[Datum]` is done to `&[T]` by a simple type casting,
+                // however, we should use `T::try_cast(&'mc PgAllocator, Datum)` to ignore nulls
+                let mem_size_datums = std::mem::size_of_val(datums);
+                let datums = if mem_size_datums == 0 {
+                    std::slice::from_raw_parts(datums.as_ptr() as *const T, 0)
+                } else {
+                    let mem_size_type = std::mem::size_of::<T>();
+                    assert_eq!(mem_size_datums % mem_size_type, 0);
+                    std::slice::from_raw_parts(
+                        datums.as_ptr() as *const T,
+                        mem_size_datums / mem_size_type,
+                    )
+                };
+
+                Ok(datums)
+            }
+        } else {
+            Err("datum was NULL")
+        }
+    }
+}
+
+impl<'mc, 's, T> From<&[T]> for PgDatum<'mc>
+where
+    'mc: 's,
+    T: 's,
+    PgDatum<'mc>: From<T>,
+{
+    fn from(_value: &[T]) -> Self {
+        // TODO
```

Comment:
so for this to work, the memory of `&[T]` needs to be `&'mc [T]`, if we want to do it with zero allocations...

Reply:
OK. I'll change the lifetime. However, for now it's a bogus implementation, as it will always return a `NULL` to PG if the Rust function return type is of `&[T]`. It still lacks the proper support for returning them